modules/pubsub/src/main/scala/com/snowplowanalytics/snowplow/streams/pubsub/PubsubFactory.scala

```diff
@@ -24,26 +24,64 @@ import java.util.concurrent.{Executors, ScheduledExecutorService}
 
 class PubsubFactory[F[_]: Async] private (
   transport: FixedTransportChannelProvider,
-  executor: FixedExecutorProvider
+  executor: FixedExecutorProvider,
+  credentials: CredentialsProvider
 ) extends Factory[F, PubsubSourceConfig, PubsubSinkConfig] {
 
   def sink(config: PubsubSinkConfig): Resource[F, Sink[F]] =
-    PubsubSink.resource(config, transport, executor)
+    PubsubSink.resource(config, transport, executor, credentials)
 
   def source(config: PubsubSourceConfig): Resource[F, SourceAndAck[F]] =
-    PubsubSource.resource(config, transport, executor)
+    PubsubSource.resource(config, transport, executor, credentials)
 }
 
 object PubsubFactory {
 
   def resource[F[_]: Async](config: PubsubFactoryConfig): Resource[F, PubsubFactory[F]] =
     for {
       executor <- executorResource[F]
-      channel <- makeChannel(config.gcpUserAgent, executor)
-    } yield new PubsubFactory(FixedTransportChannelProvider.create(channel), FixedExecutorProvider.create(executor))
+      credentialsProvider <- Resource.eval(makeCredentialsProvider[F](config))
+      channel <- makeChannel(config, executor)
+    } yield new PubsubFactory(FixedTransportChannelProvider.create(channel), FixedExecutorProvider.create(executor), credentialsProvider)
 
-  def makeChannel[F[_]: Sync](gcpUserAgent: GcpUserAgent, executor: ScheduledExecutorService): Resource[F, TransportChannel] = {
-    def withCredentials(credentials: Credentials) = Sync[F].delay {
+  def makeCredentialsProvider[F[_]: Sync](config: PubsubFactoryConfig): F[CredentialsProvider] =
+    config.emulatorHost match {
+      case Some(_) =>
+        // The pubsub emulator is enabled, so do not try to resolve credentials
+        Sync[F].pure(NoCredentialsProvider.create())
+      case None =>
+        // Pubsub emulator _not_ enabled. Use real credentials for real pubsub
+        Sync[F]
+          .blocking(TopicAdminSettings.defaultCredentialsProviderBuilder().build().getCredentials())
+          .map(FixedCredentialsProvider.create(_))
```

Comment:
Is this not all automated when you set `PUBSUB_EMULATOR_HOST` like [here](https://github.com/snowplow/stream-collector/blob/3.4.0/pubsub/src/it/scala/com/snowplowanalytics/snowplow/collectors/scalastream/it/pubsub/Containers.scala#L65) ? I don't remember adding something special for the emulator, but that might be because we are not defining the channel manually

Reply:
The `PUBSUB_EMULATOR_HOST` option you see in the collector only works because it is read back in from the environment [here](https://github.com/snowplow/stream-collector/blob/3.4.0/pubsub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/BuilderOps.scala#L29-L30).  It is not automatic.  The collector still needs to explicitly set the host on the GRPC channel builder.

Reply:
Oops, my mistake!